### PR TITLE
Add support to command liveness probe

### PR DESCRIPTION
**Pull request by lonefreak**
The changes will allow to add one command liveness probe to a container. You can specify:
- command to execute (does not suppot httpGet) - required field
- timeout seconds
- initial delay seconds
- failure threshold
- period seconds
- success threshold

### CODE REVIEW

**Comment by carlossg** [comment]
Can you add some text to the readme to show the feature and the help-*.html files like https://github.com/lonefreak/kubernetes-plugin/tree/7dd3067c2d5707c8ed4ada614c046448d0d6d11a/src/main/resources/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplate

It doesn't provide pipeline support?

**Comment by lonefreak** [comment]
I will add info to README and the help files.

As for the pipeline support, I didn't do that. I can probably do it quickly. Would you like to have that in this PR or can I add pipeline support in a different one?

**Comment by carlossg** [comment]
here would be better, thanks!

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplate.java [hunk 1]
```diff
@@ -150,6 +152,13 @@ public void setEnvVars(List<ContainerEnvVar> envVars) {
         this.envVars.addAll(envVars);
     }
 
+    public List<ContainerLivenessProbe> getLivenessProbe() { return livenessProbe; }
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplate.java]: livenessProbe will be null when upgrading from a previous version
You can either check for it and return emptyList or implement readResolve as in https://github.com/jenkinsci/kubernetes-plugin/blob/6b35af2aeebdf81da76a705d1ead94010437c594/src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java#L446

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplate.java [hunk 2]
```diff
@@ -46,6 +46,8 @@
 
     private final List<ContainerEnvVar> envVars = new ArrayList<ContainerEnvVar>();
 
+    private final List<ContainerLivenessProbe> livenessProbe = new ArrayList<ContainerLivenessProbe>();
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplate.java]: do we need a list ? seems that you can only have one

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 3]
```diff
@@ -379,6 +370,20 @@ private Container createContainer(KubernetesSlave slave, ContainerTemplate conta
             containerMounts.add(new VolumeMount(containerTemplate.getWorkingDir(), WORKSPACE_VOLUME_NAME, false, null));
         }
 
+        Probe livenessProbe = null;
+        for (ContainerLivenessProbe clp : containerTemplate.getLivenessProbe()) {
+            if (parseLivenessProbe(clp.getExecArgs()) != null) {
+                livenessProbe = new ProbeBuilder()
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: you have a list of probes and then you just overwrite all of them but the latest one
should containerTemplate just have one `livenessProbe` ?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 4]
```diff
@@ -535,6 +541,25 @@ private Pod getPodTemplate(KubernetesSlave slave, @CheckForNull Label label) {
...
     }
 
+    /**
+     * Split a command in the parts that LivenessProbe need
+     * 
+     * @param livenessProbeExec
+     * @return
+     */
+    List<String> parseLivenessProbe(String livenessProbeExec) {
+        if (livenessProbeExec == null || livenessProbeExec.isEmpty()) {
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: there's `StringUtils.isBlank(livenessProbeExec)` if you want to use it

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 5]
```diff
@@ -535,6 +541,25 @@ private Pod getPodTemplate(KubernetesSlave slave, @CheckForNull Label label) {
...
+            return null;
+        }
+        // handle quoted arguments
+        Matcher m = SPLIT_IN_SPACES.matcher(livenessProbeExec);
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: parsing is typically a source for problems, either make exec be a list in the UI , or add some good unit tests with multiple arguments (quoted, unquoted, with spaces, single quotes,...)

Comment (lonefreak) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: Yeah, I should have done that. Completely forgot in the end. Will do it.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 3 above)
Comment (lonefreak) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: Yes, you are correct. I will fix this.

**Comment by lonefreak** [comment]
Took me too long to make all the changes and the plugin has evolved a lot. I will close this PR and open a new one based on the latest state.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplate.java (hunk 2 above)
Comment (lonefreak) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplate.java]: Done in #158 .

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplate.java (hunk 1 above)
Comment (lonefreak) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplate.java]: Done in #158

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 3 above)
Comment (lonefreak) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: Fixed in #158

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 4 above)
Comment (lonefreak) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: Done in #158

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 5 above)
Comment (lonefreak) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: Added some tests to fix this in #158

**Pull request closed without merging**